src/main/scala/shine/DPIA/fromRise.scala

```diff
@@ -673,31 +673,32 @@ object fromRise {
             Let(s, t, a, x, f)))
       }
 
-      case r.ForeignFunction(decl) =>
-        def collectTypes(t: PhraseType): (Vector[DataType], DataType) = {
+      case core.foreignFunction(decl, n) =>
+        def collectTypes(t: PhraseType): (Seq[DataTypeIdentifier], DataTypeIdentifier) = {
           t match {
-            case ExpType(dt: DataType, `read`) =>
-              (Vector(), dt)
-            case FunType(ExpType(dt: DataType, `read`), out) =>
+            case ExpType(dt: DataTypeIdentifier, `read`) => (Seq(), dt)
+            case FunType(ExpType(dt: DataTypeIdentifier, `read`), out) =>
               val (i, o) = collectTypes(out)
               (dt +: i, o)
-            case _ =>
-              throw new Exception("This should not be possible")
+            case DepFunType(_, t) => collectTypes(t)
+            case _ => throw new Exception("This should not be possible")
           }
         }
         val (inTs, outT) = collectTypes(t)
-
-        def buildFFCall(args: Vector[Phrase[ExpType]]
-                            ): Phrase[_ <: PhraseType] = {
-          val i = args.length
-          if (i < inTs.length) {
-            fun[ExpType](ExpType(inTs(i), read), a =>
-              buildFFCall(args :+ a))
-          } else {
-            ForeignFunctionCall(decl, inTs, args)(outT)
-          }
+        assert(inTs.length == n)
+
+        inTs.foldRight[Phrase[_ <: PhraseType]](
+          depFun[DataKind](outT)({
+            val args = Seq.tabulate(n)(i => Identifier(freshName("x"), ExpType(inTs(i), read)))
```

Comment:
The `fun` syntax creates a scala binding that has to be scoped appropriately and I think this would have made the implementation much more complicated.